### ISSUE TITLE
Authorize.net issues on magento 2.2.0 version  main.CRITICAL: Gateway error: This transaction has been declined. {"exception":"[object] (Magento\\Framework\\Exception\\LocalizedException(code: 0): Gateway error: This transaction has been declined. at /vendor/magento/module-authorizenet/Model/Directpost.php:635)"} []

### ISSUE THREAD

**Issue by vijaymagento1** [magento/magento2#12311]
When we place the order in magento 2.2.1 version by using the authorize.net payment live credentials then it give the error message "An error occurred on the server, Please try to place order again."
And internally placed the order and i can see in the admin as well as in authorize.net

**Following errors on the exception log** 

main.CRITICAL: Gateway error: This transaction has been declined. {"exception":"[object] (Magento\\Framework\\Exception\\LocalizedException(code: 0): Gateway error: This transaction has been declined. at /vendor/magento/module-authorizenet/Model/Directpost.php:635)"} []

[2017-11-16 08:58:08] main.CRITICAL: The transaction was declined because the response hash validation failed. {"exception":"[object] (Magento\\Framework\\Exception\\LocalizedException(code: 0): The transaction was declined because the response hash validation failed. at /vendor/magento/module-authorizenet/Model/Directpost.php:539)"} []

**Following errors on the Debug.log**

[2017-11-16 08:54:12] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["interception","CONFIG"],"mode":"matchingTag"}} []
[2017-11-16 08:54:12] main.DEBUG: cache_invalidate:  {"method":"GET","url":"https://xxxxx.com/admin/admin/cache/index/key/455sds5c5s8d5sd5s8s8d8sd8s8d8s8d5sd8s8d5ds8/","invalidateInfo":{"tags":["interception","CONFIG"],"mode":"matchingTag"}} []

Please review the issues and let us know ASAP.

Thanks!

**Comment by YevSent** [comment]
@vijaymagento1, please, enable debug mode for authorize.net payment method and provide transaction  details from debug.log

Also, if it possible, please, provide a response from transact.dll, you can in a browser console log during place order action.

**Comment by vijaymagento1** [comment]
Hi Joni,
We have already enabled the debug mode for the authorize.net payment method and there are the following log which we have got and in the console we didn't get any thing and in the backend we can see the placed order but in the frontend showing error **"An error occurred on the server, Please try to place order again."** and not redirecting to success page.

**Following errors on the Debug.log**

[2017-11-16 08:54:12] main.DEBUG: cache_invalidate: {"method":"GET","url":"http:/","invalidateInfo":{"tags":["interception","CONFIG"],"mode":"matchingTag"}} []
[2017-11-16 08:54:12] main.DEBUG: cache_invalidate: {"method":"GET","url":"https://xxxxx.com/admin/admin/cache/index/key/455sds5c5s8d5sd5s8s8d8sd8s8d8s8d5sd8s8d5ds8/","invalidateInfo":{"tags":["interception","CONFIG"],"mode":"matchingTag"}} []


Thanks!

**Comment by YevSent** [comment]
Your debug log doesn't contain any transaction details, I suppose your instance doesn't work properly. The debug.log should contain transaction request and response. Did you enabled debug mode globally for all Magento application in *Stores*->*Configuration*->*Advanced*->*Developer*->*Debug*->*Log to file* (option available only in developer mode)?

**Comment by OlgaBurtyka** [comment]
@joni-jones same issue
![issue](https://user-images.githubusercontent.com/4771407/33165272-dd052582-d03e-11e7-8cf5-a85222ef44f2.png)
![response](https://user-images.githubusercontent.com/4771407/33165273-dd21eadc-d03e-11e7-8edc-a33566c8b269.png)

**Comment by YevSent** [comment]
@OlgaBurtyka, from https://developer.authorize.net/api/reference/responseCodes.html documentation
`x_response_code` => 2 does not contain any additional information. I suggest you can find more details in your Authorize.net merchant account.

Also, please, attach your debug log for the transaction request, maybe, it will help to clarify an issue.

**Comment by OlgaBurtyka** [comment]
@joni-jones in auth account not any info about transactions decline.

**Comment by OlgaBurtyka** [comment]
It difficult to reproduce, some order placed via Authorize - some declined, but after several attempts the transaction is created.  For example customer placed order - and got 2 decline transactions and 1 success. And as result in admin we have 2 canceled orders and one in progress. It start happen after update to 2.2.1 version

**Comment by YevSent** [comment]
Is it sandbox or production account? Without logs, it's hard to understand what's going wrong.

**Comment by magento-engcom-team** [comment]
@vijaymagento1, we are closing this issue due to inactivity. If you'd like to update it, please reopen the issue.

**Comment by AirmanAJK** [comment]
\vendor\magento\module-authorizenet\Model\Directpost.php has a method validateResponse that compares the hash value returned by Authorize.NET with the computed hash of the response values. If they don't match, an exception is thrown. This is for security purpose, so that it knows the response has actually come from Authorize.NET. You probably have an incorrect Authorize.net hash value configured in Magento. I believe you can regenerated one with Authorize.net if you need to. Good luck.

**Comment by devcflynn** [comment]
We are having this same issue. We have followed all the same steps, making sure sandbox is off, debug was set to "true". Debug logs all say the transaction was successful. We get the charge in Authorize.net and everything looks good. The issue still persists and I have also confirmed that we are indeed using the MD5 Hash that is configured in Authorize.net in our settings as well. 

When a customer checks out, it says there was an error, but it still creates the order in Magento and Authroize.net. Below is a debug output from the last transaction we did. The customer data has been redacted for security, but this is the debug response we are getting from these transactions. 

Magento ver. 2.1.10

Client side: "An error occurred on the server, Please try to place order again."

Error logs: main.CRITICAL: The transaction was declined because the response hash validation failed. {"exception":"[object] (Magento\Framework\Exception\LocalizedException(code: 0): The transaction was declined because the response hash validation failed. at /vendor/magento/module-authorizenet/Model/Directpost.php:539)"} []

[2017-12-07 15:17:39] main.DEBUG: array (
  'response' => 
  array (
    'x_response_code' => '1',
    'x_response_reason_code' => '1',
    'x_response_reason_text' => 'This transaction has been approved.',
    'x_avs_code' => 'Y',
    'x_auth_code' => 'xxxxxxx',
    'x_trans_id' => 'xxxxxxxxx',
    'x_method' => 'CC',
    'x_card_type' => 'American Express',
    'x_account_number' => 'XXXXXXX',
    'x_first_name' => 'xxxxxxx',
    'x_last_name' => 'xxxxxxx',
    'x_company' => 'xxxxxxxxxxxxxx',
    'x_address' => 'xxxxxxxxxx',
    'x_city' => 'xxxxxx',
    'x_state' => 'xxxxxx',
    'x_zip' => 'xxxxxx',
    'x_country' => '',
    'x_phone' => 'xxxxxxxx',
    'x_fax' => '',
    'x_email' => 'xxxxxxxxx',
    'x_invoice_num' => '000000015',
    'x_description' => '',
    'x_type' => 'auth_capture',
    'x_cust_id' => '',
    'x_ship_to_first_name' => 'xxxxxxx',
    'x_ship_to_last_name' => 'xxxxxxx',
    'x_ship_to_company' => 'xxxxxxx',
    'x_ship_to_address' => 'xxxxxx',
    'x_ship_to_city' => 'xxxxxx',
    'x_ship_to_state' => 'xxxxxxxxx',
    'x_ship_to_zip' => 'xxxxxxxx',
    'x_ship_to_country' => '',
    'x_amount' => '7.12',
    'x_tax' => '0.12',
    'x_duty' => '0.00',
    'x_freight' => '5.00',
    'x_tax_exempt' => 'FALSE',
    'x_po_num' => '',
    'x_MD5_Hash' => 'xxxxxxxxxxxxxxxxxxxxxxxxxxx',
    'x_SHA2_Hash' => '',
    'x_cvv2_resp_code' => 'M',
    'x_cavv_response' => '',
    'x_test_request' => 'false',
    'controller_action_name' => 'directpost_payment',
    'is_secure' => '1',
  ),
) {"is_exception":false} []

**Comment by AirmanAJK** [comment]
How have you confirmed you are using the correct hash? Authorize.NET never shows you the hash value again once it is set. All you can do is regenerate a new hash value, and update Magento with that exact new value. You almost certainly have the wrong value set.

The code in Magento for this is pretty straightforward and is unlikely a bug:

`return strtoupper(md5($merchantMd5 . $merchantApiLogin . $transactionId . $amount));`

If you want to confirm this, generate the hash yourself in PHP using the above method and compare it with the x_MD5_Hash from the Authorize.NET response. If they don't match, you are configured wrong.

**Comment by devcflynn** [comment]
Hey @AirmanAJK ,
I was 5 min late to updating this issue before you jumped in, thanks for responding. After hours of digging, I have discovered that the problem was the MD5 hash I was using, but not because they were mismatched. Turns out that no where in any docs (besides a comment left deep in another unrelated issue somewhere else) does it mention that the **hash should only be a maximum of 10 characters**. Turns out, that was the only problem. When I updated the hash on both Auth.net and Magento to be only 10 characters, it fixed my issue.

It would be nice if this was mentioned in the docs some where. I imagine this may confuse other people using Magento for the first time. But just a thought.... If someone has docs related to what I just mentioned. Probably not a terrible idea to link it here or somewhere searchable. 

Thanks for the help!

**Comment by AirmanAJK** [comment]
Strange. I found a few results on Google that also fixed their issue with a 10 character hash... strange. I looked closer at the code and there doesn't appear to be anything that would cause this. No string trimming or weird manipulation. Authorize.NET supports up to 20 characters for the MD5 hash, but I guess somehow Magento ends up with a different hash of the same values...? No idea. Glad your issue is fixed.
